File: app/src/main/java/com/websarva/wings/android/zuboradiary/data/diary/ItemNumber.java
package com.websarva.wings.android.zuboradiary.data.diary;

import androidx.annotation.NonNull;

import java.io.Serializable;
import java.util.Objects;

public class ItemNumber implements Serializable, Comparable<ItemNumber> {

    private final int value;
    public static final int MIN_NUMBER = 1;
    public static final int MAX_NUMBER = 5;

    public ItemNumber(int value) {
        if (value < MIN_NUMBER || value > MAX_NUMBER) throw new IllegalArgumentException();

        this.value = value;
    }

    public ItemNumber incrementNumber() {
        return new ItemNumber(value + 1);
    }

    public ItemNumber decrementNumber() {
        return new ItemNumber(value - 1);
    }

    public int getValue() {
        return value;
    }

    @NonNull
    @Override
    public String toString() {
        return String.valueOf(value);
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) return false;
        if (!(obj instanceof ItemNumber)) return false;
        ItemNumber target = (ItemNumber) obj;
        return this.getValue() == target.getValue();
    }

    @Override
    public int compareTo(ItemNumber itemNumber) {
        Objects.requireNonNull(itemNumber);
        return Integer.compare(this.value, itemNumber.value);
    }
}

File: app/src/main/java/com/websarva/wings/android/zuboradiary/data/model/ItemNumber.kt
package com.websarva.wings.android.zuboradiary.data.model

import java.io.Serializable

internal class ItemNumber(value: Int) : Serializable, Comparable<ItemNumber> {

    companion object {
        const val MIN_NUMBER: Int = 1
        const val MAX_NUMBER: Int = 5
    }

    val value: Int

    init {
        require(value >= MIN_NUMBER)
        require(value <= MAX_NUMBER)

        this.value = value
    }

    fun inc(): ItemNumber {
        val value = this.value.inc()
        require(value <= MAX_NUMBER)

        return ItemNumber(value)
    }

    override fun toString(): String {
        return value.toString()
    }

    override fun equals(other: Any?): Boolean {
        if (other == null) return false
        if (other !is ItemNumber) return false
        return this.value == other.value
    }

    override fun compareTo(other: ItemNumber): Int {
        return this.value.compareTo(other.value)
    }

    override fun hashCode(): Int {
        return value
    }
}
